validator/src/journal/candidate_block.rs

```diff
@@ -220,4 +227,180 @@ impl CandidateBlock {
+                .call_method(
+                    py,
+                    "check_publish_block",
+                    (self.block_builder
+                        .getattr(py, "block_header")
+                        .expect("BlockBuilder has no attribute 'block_header'"),),
+                    None,
+                )
+                .expect("consensus has no method 'check_publish_block'")
+                .extract::<bool>(py)
+                .unwrap()
+        } {
+            return Err(BlockPublisherError::ConsensusNotReady);
+        }
+
+        self.scheduler.finalize(true).unwrap();
+        let execution_results = self.scheduler.complete(true).unwrap().unwrap();
+
+        let mut committed_txn_cache = {
+            let py = unsafe { cpython::Python::assume_gil_acquired() };
+            TransactionCommitCache::new(self.block_store.clone_ref(py))
+        };
+
+        let batches_w_no_results: Vec<String> = execution_results
+            .batch_results
+            .iter()
+            .filter(|(_, txns)| txns.is_none())
+            .map(|(batch_id, _)| batch_id.clone())
+            .collect();
+
+        let valid_batch_ids: Vec<String> = execution_results
+            .batch_results
+            .into_iter()
+            .filter(|(_, txns)| match txns {
+                Some(t) => !t.iter().any(|t| !t.is_valid),
+                None => false,
+            })
+            .map(|(b_id, _)| b_id)
+            .collect();
+
+        let builder = {
+            let py = unsafe { cpython::Python::assume_gil_acquired() };
+            self.block_builder.clone_ref(py)
+        };
+        let mut bad_batches = vec![];
+        let mut pending_batches = vec![];
+
+        for batch in self.pending_batches.clone() {
+            let header_signature = &batch.header_signature.clone();
+            if batch.trace {
+                debug! {
+                    "TRACE {} : CandidateBlock finalize",
+                    header_signature
+                }
+            }
+
+            if batches_w_no_results.contains(&batch.header_signature) {
+                if !self.injected_batch_ids
+                    .contains(batch.header_signature.as_str())
+                {
+                    pending_batches.push(batch)
+                } else {
+                    warn! {
+                        "Failed to inject batch {}",
+                        header_signature
+                    };
+                }
+            } else if valid_batch_ids.contains(&batch.header_signature) {
+                if !self.check_batch_dependencies(&batch) {
+                    debug! {
+                        "Batch {} is invalid, due to missing txn dependency",
+                        header_signature
+                    };
+                    bad_batches.push(batch.clone());
+                    pending_batches.clear();
+                    pending_batches.append(&mut self.pending_batches
+                        .clone()
```

Comment:
If you `std::mem::replace` this probably becomes easier too (doesn't require clone())